File: 201902715/10976.cc
#include <iostream>
#include <algorithm>
#include <vector>
#include <cstring>
#include <queue>

using namespace std;

typedef long long ll;

int num_nodes, num_edges;
vector<int> graph[201];
int capacity[201][201], flow[201][201];
int previous[201];

int find_maximum_flow() {
    int result = 0;
    while (true) {
        queue<int> q;
        q.push(1);
        memset(previous, -1, sizeof(previous));
        previous[1] = 0;
        while (!q.empty()) {
            int current = q.front();
            q.pop();
            for (int next : graph[current]) {
                if (capacity[current][next] - flow[current][next] > 0 && previous[next] == -1) {
                    previous[next] = current;
                    q.push(next);
                    if (next == num_nodes) { break; }
                }
            }
        }
        if (previous[num_nodes] == -1) { break; }
        int current_flow = 999999999;
        for (int i = num_nodes; i != 1; i = previous[i]) {
            current_flow = min(current_flow, capacity[previous[i]][i] - flow[previous[i]][i]);
        }
        for (int i = num_nodes; i != 1; i = previous[i]) {
            capacity[previous[i]][i] -= current_flow;
            capacity[i][previous[i]] += current_flow;
        }
        result += current_flow;
    }
    return result;
}

int main() {
    int num_tests;
    cin >> num_tests;
    while (num_tests--) {
        cin >> num_nodes >> num_edges;
        for (int i = 1; i <= num_edges; i++) {
            int start, end;
            cin >> start >> end;
            if (start == 1 || end == num_nodes) { capacity[start][end] = 1; }
            else { capacity[start][end] = 999999999; }
            graph[start].push_back(end);
            graph[end].push_back(start);
        }
        cout << find_maximum_flow() << endl;
        for (int i = 1; i <= num_nodes; i++) {
            graph[i].clear();
        }
        memset(capacity, 0, sizeof(capacity));
        memset(flow, 0, sizeof(flow));
    }
    return 0;
}

File: 201902715/10976.py
from collections import deque
from typing import List
import sys

def find_maximum_flow(num_nodes: int, graph: List[List[int]], capacity: List[List[int]], flow: List[List[int]]) -> int:
    result = 0
    while True:
        q = deque([1])
        previous = [-1] * (num_nodes + 1)
        previous[1] = 0

        while q:
            current = q.popleft()
            for next_node in graph[current]:
                if capacity[current][next_node] - flow[current][next_node] > 0 and previous[next_node] == -1:
                    previous[next_node] = current
                    q.append(next_node)
                    if next_node == num_nodes:
                        break
        
        if previous[num_nodes] == -1:
            break

        current_flow = sys.maxsize
        i = num_nodes
        while i != 1:
            current_flow = min(current_flow, capacity[previous[i]][i] - flow[previous[i]][i])
            i = previous[i]

        i = num_nodes
        while i != 1:
            capacity[previous[i]][i] -= current_flow
            capacity[i][previous[i]] += current_flow
            i = previous[i]

        result += current_flow

    return result


def main():
    num_tests = int(input())
    for _ in range(num_tests):
        num_nodes, num_edges = map(int, input().split())
        graph = [list() for _ in range(num_nodes + 1)]
        capacity = [[0] * (num_nodes + 1) for _ in range(num_nodes + 1)]
        flow = [[0] * (num_nodes + 1) for _ in range(num_nodes + 1)]

        for _ in range(num_edges):
            start, end = map(int, input().split())
            if start == 1 or end == num_nodes:
                capacity[start][end] = 1
            else:
                capacity[start][end] = sys.maxsize
            
            graph[start].append(end)
            graph[end].append(start)

        print(find_maximum_flow(num_nodes, graph, capacity, flow))

        for i in range(1, num_nodes + 1):
            graph[i].clear()

if __name__ == "__main__":
    main()
